src/NBitcoin/Network.cs

```diff
@@ -230,6 +230,102 @@ internal static Network Register(Network network)
+        /// Console.WriteLine("time: " + header.Time);
+        /// Console.WriteLine("version: " + header.Version);
+        /// Console.WriteLine("hash: " + header.GetHash());
+        /// Console.WriteLine("merkleroot: " + header.HashMerkleRoot);
+        /// </code>
+        /// <returns> A genesis block. </returns>
+        public static Block MineGenesisBlock(ConsensusFactory consensusFactory, string pszTimestamp, Target target, Money genesisReward, int version = 1)
+        {
+            if (target == null)
+                throw new ArgumentException("Parameter 'target' cannot be null. Example use: new Target(new uint256(\"0000ffff00000000000000000000000000000000000000000000000000000000\"))");
+
+            if (consensusFactory == null)
+                throw new ArgumentException("Parameter 'consensusFactory' cannot be null. Use 'new ConsensusFactory()' for Bitcoin-like proof-of-work blockchains" +
+                                            "and 'new PosConsensusFactory()' for Stratis-like proof-of-stake blockchains.");
+
+            if (string.IsNullOrEmpty(pszTimestamp))
+                throw new ArgumentException("Parameter 'pszTimestamp' cannot be null. Use a news headline or any other appropriate string.");
+
+            if (genesisReward == null)
+                throw new ArgumentException("Parameter 'genesisReward' cannot be null. Example use: 'Money.Coins(50m)'.");
+
+            DateTimeOffset time = DateTimeOffset.Now;
+            uint unixTime = Utils.DateTimeToUnixTime(time);
+
+            Transaction txNew = consensusFactory.CreateTransaction();
+            txNew.Version = (uint)version;
+            txNew.Time = unixTime;
+            txNew.AddInput(new TxIn()
+            {
+                ScriptSig = new Script(
+                    Op.GetPushOp(0),
+                    new Op()
+                    {
+                        Code = (OpcodeType)0x1,
+                        PushData = new[] { (byte)42 }
+                    },
+                    Op.GetPushOp(Encoders.ASCII.DecodeData(pszTimestamp)))
+            });
+
+            txNew.AddOutput(new TxOut()
+            {
+                Value = genesisReward,
+            });
+
+            Block genesis = consensusFactory.CreateBlock();
+            genesis.Header.BlockTime = time;
+            genesis.Header.Bits = target;
+            genesis.Header.Nonce = 0;
+            genesis.Header.Version = version;
+            genesis.Transactions.Add(txNew);
+            genesis.Header.HashPrevBlock = uint256.Zero;
+            genesis.UpdateMerkleRoot();
+
+            // Iterate over the nonce until the block header hash is under the target.
+            uint256 hashTarget = target.ToUInt256();
+            while (genesis.Header.GetHash() >= hashTarget)
+            {
+                genesis.Header.Nonce++;
+                if (genesis.Header.Nonce == 0)
+                    genesis.Header.Time++;
+            }
```

Comment:
It might be a good idea if the nonce value comes as an optional argument and we do this hash finding loop only if it is not given – i.e. it has default value of 0. Such argument would allow us to precompute the correct nonce in the way that the creation of the Genesis can then be done instantly.

Reply:
So this method is to be executed on its own, just to mine the genesis block.
Then, the creator of the network should look at the data included in this block and use it in the network settings.
As such, it's not a replacement for the CreateGenesisBlock methods, whose signature is `Block CreateStratisGenesisBlock(ConsensusFactory consensusFactory, uint nTime, uint nNonce, uint nBits, int nVersion, Money genesisReward)`.

Now thinking about it, I could add nonce and time as optional params.
I'll try to see how it looks.

Reply:
I leave this decision to you, it was just an idea, maybe use it in different way than I suggested.